src/UserManager.ts

```diff
@@ -263,15 +263,42 @@ export class UserManager {
     }
 
     protected async _useRefreshToken(state: RefreshState): Promise<User> {
-        const response = await this._client.useRefreshToken({
-            state,
-            timeoutInSeconds: this.settings.silentRequestTimeoutInSeconds,
+        const refreshUser = async (): Promise<User> => {
+            const response = await this._client.useRefreshToken({
+                state,
+                timeoutInSeconds: this.settings.silentRequestTimeoutInSeconds,
+            });
+            return new User({ ...state, ...response });
+        };
+
+        if (!navigator.locks) {
+            // Legacy option for older browser which don't support `navigator.locks`.
+            const user = await refreshUser();
+            await this.storeUser(user);
+            this._events.load(user);
+            return user;
+        }
+
+        const broadcastChannel = new BroadcastChannel(`refresh_token_${state.refresh_token}`);
+        let user : User | null = null;
+
+        broadcastChannel.addEventListener("message", (event : MessageEvent<User>) => {
+            user = event.data;
         });
-        const user = new User({ ...state, ...response });
 
-        await this.storeUser(user);
-        this._events.load(user);
-        return user;
+        return await navigator.locks.request(
+            `refresh_token_${state.refresh_token}`,
+            async () => {
+                if (!user) {
+                    user = await refreshUser();
+                    broadcastChannel.postMessage(user);
+                }
```

Comment:
I like this implementation best.

@kherock, just a few questions:
1. Why is `storeUser(user)` only performed by the leader ? (and not by everyone at the end, like `_events.load`)
2. Is a timeout necessary in case `postMessage` never occurred ? (ex: `refreshUser()` taking too long and in the meantime user closes the leader tab) 
3. Still wondering about the proper lock key to use... 🤔

Reply:
I didn't think about the scenario of the tab being interrupted, but more realistically, my code doesn't handle when the leader encounters an error either.

`storeUser` writes the user to `localStorage`/`sessionStorage` which is presumably shared by both tabs in this scenario, so it only needs to be serialized and stored once. The user actually doesn't even need to be sent over the channel at all, the other tab could just as well do `return await this._loadUser()` after receiving the notification from the leader (or by waiting for the lock to release, though this doesn't account for the lock releasing because of a failure or interruption).

I think think finding a "proper" lock key to use in a separate issue. My hunch is that we should lock onto `session_state` when the IDP implements OIDC sessions (e.g. Keycloak) and fall back to the refresh token.

Reply:
- Agreed that some error handling needs to be made.

- > which is presumably shared by both tabs in this scenario

  If I'm not mistaken, `sessionStorage` is not shared between tabs - so it won't work in that case.

- I've seen `session_state` being returned by Keycloak (but not Auth0), so yes we could use it if we have it or fall back otherwise. But I'm still not confident in using `refresh token` as the fall back. Each tab receives a different refresh token, which makes the purpose of the lock completely useless.

Reply:
If sessionStorage isn't shared, how would two tabs obtain the same refresh token in the first place?

Reply:
I don't know.. I don't think we ever talked about that scenario.. 👀

To me the issue here is about concurrency.
Two tabs, same client_id, the first tab to refresh the tokens wins and kills the others due to refresh token rotation.

Reply:
Yes, but as I explained in that comment I linked, having the same client ID should not imply that the same session is in use. Refresh token rotation is tied to a particular user's session, not a client altogether. Perhaps the client ID plus the `sub` claim is more appropriate?

Reply:
Yes, I think so. Good idea

Reply:
@kherock What I'm doing is already leader election. What your implementation is missing is that if the leader dies throughout the request, the other tabs will never get a response.

Reply:
@Badisi Actually, the refresh token is being shared if both tabs originated from the same login (localStorage). In that case, the refresh token will always be the same for both.

SessionStorage is a different topic though, but that one is not covered by this PR, as it involves a different kind of tab linking.

Reply:
@kherock client ID plus subject could work as an identifier. At least for a first spot, I don't see how that would break.

Reply:
@DASPRiD I see. Do you think that instead of sending the user over the broadcast channel, you could just check the storage and see if the refresh token changed? That way it eliminates the potential race condition between obtaining the lock and receiving the BroadcastChannel message.

```suggestion
                user = await this._loadUser();
                if (user.refresh_token === state.refresh_token) {
                  user = await refreshUser();
                }
```

Reply:
Well, assuming that the storage is synchronous and has consistent read (I don't see why it shouldn't), this would be a valid way to solve this, at least for LocalStorage. 

This would of course not apply to SessionStorage, as two tabs might just use a duplicated, but not shared, SessionStorage. But considering that this PR is not there to solve the SessionStorage duplication issue, that might be just fine. It's a pretty extreme edge case anway, I don't think that the average user duplicates their tabs :)

Reply:
@DASPRiD, I'm using neither LocalStorage nor SessionStorage, I'm keeping everything in memory. But as I see it, this concurrency issue exists no matter the type of storage, so I would opt for a way to solve it once and for all (and not only for LocalStorage). Why do you think your proposal does only cover LocalStorage ? I was convinced that simply using a common lock key would be enough to solved it for SessionStorage/MemoryStorage too. Don't you think ?

Reply:
Actually no, this issue does not affect memory storage, as memory storage is always bound to it's own session without sharing it with any other process.

The problem with SessionStorage is the part about session storage duplication when a tab is duplicated. In that instance, they will share the same session (from the perspective of the IDP), but not from their own perspective.

Reply:
@kherock Actually, I do see an issue with using client ID + sub as the lock key: It would block unrelated tabs with different sessions but same client ID + sub from performing their refresh.

Reply:
Ok, then we are talking about two different concurrency issues.
- The one you are facing, ie. when tokens are physically shared between tabs (over a storage). Tabs should wait for the first one to refresh it for the others.
- And the one I'm facing (only seen with Keycloak and might be a bug but didn't get any answers from them yet), where the first tab to do a refresh will cancel any previous generated tokens for that same client_id/session (ie. the ones generated from the other tabs). So here, tabs should also wait for the first one to refresh the tokens and send them to the others. This feels hacky and weird (because they started with different tokens and ends with identical ones) but at least your PR could solve it, as long as the lock key is identical for each tabs 🙂.

Reply:
Well yeah, the issue is absolutely on their side. Solving it with a shared lock key like that is a hacky way to solve your issue though. Especially since it would only really solve it as long as your are using local storage. If you are using anything else (based on Kheros asked change of not using the BroadcastChannel), this would not help with your issue.

It would, as mentioned above, also introduce the issue that it would block possibly unrelated tabs from performing their refresh. I really only see one way here to address your issue without affecting all other users:

1. You would have to opt into local storage (since Keycloak forces a shared session on you anyway).
2. Add an option to the Usermanager so you can opt into using client ID + sub as the lock key, although the default would be the refresh token.

@kherock Actually, instead of using the refresh token, we could also add a randomly generated ID to the store which is used as common identifier for the lock, but when `session_state` is available, store that as the common identifier. That way it would address @Badisi issue, and make us not rely on the value of the refresh token.

Although to ultimately address  their issue, we'd need to use the BroadcastChannel. Thinking about that, we could add a delay of maybe 250ms or so into the leader function. That would address any possible common race condition. We probably might not even need such a long delay.